pip/pip-302.md

```diff
@@ -0,0 +1,104 @@
+
+For more detailed information about the TableView, please refer to the [Pulsar documentation](https://pulsar.apache.org/docs/next/concepts-clients/#tableview).
+
+# Motivation
+
+In the context of utilizing the TableView component, there are instances where we aspire to consistently retrieve the most up-to-date value associated with a given key. 
+When read and write operations for a certain key do not occur simultaneously, we can refresh the TableView before reading the key to obtain the latest value for this key.
+To accomplish this, we can employ an API `refreshAsync` that allows us to refresh the TableView before accessing the value corresponding to the desired key.
+# Goals
+
+## In Scope
+
+The proposed changes aim to add a new API that triggers the reading of all existing messages that before the trigger time, 
+updates the TableView, and waits for the read operation to complete. This will be achieved by introducing the `refreshAsync()` method.
+
+## Out of Scope
+
+No out-of-scope items have been identified at this time.
+
+# High-Level Design
+
+The high-level design involves the addition of a new API method called `refreshAsync()`. 
+This method will be responsible for initiating the process of reading all existing messages that before the trigger time,
+updating the TableView, and returning a future to wait for the read operation to complete.
+
+# Detailed Design
+
+## Design & Implementation Details
+
+The `refreshAsync()` method will be implemented as follows:
+
+```java
+@Override
+public CompletableFuture<Void> refreshAsync() {
+        return reader.thenCompose(this::readAllExistingMessages);
+}
+```
+
+# Public-Facing Changes
+
+## Public API
+
+The following changes will be made to the public API:
+
+### `refreshAsync()`
+
+This new API method triggers the reading of all existing messages, updates the TableView, and waits for the read operation to complete.
+```java
+/**
+ * Triggers the reading of all existing messages from the topics, updates the TableView and waits for the read operation to complete.
+ * This method fetches the last message of the topics at the point of invocation and updates the TableView with all messages up to and including this last message.
+ * After the update is complete, users can use the TableView to obtain the latest value for any key. 
+ * Note that the 'latest' value refers to the value at the point of calling refresh, not necessarily the current latest if more messages have been produced in the meantime.
+ *
+ * @return a CompletableFuture that completes when all existing messages up to the point of invocation have been read, and the TableView has been updated.
+ *
+ * Example usage:
+ * table.refreshAsync().whenComplete((__, ex) -> {
+ *     if (ex == null) {
+ *         table.get(key);
+ *     }
+ * });
```

Comment:
```suggestion
 * table.refreshAsync().thenApply(__ -> table.get(key));
```

Make the code example more meaningful. Maybe you can add the comment that use the code above instead of the trivial `table.get(key)`.

Actually, I think it's not much different from a `getAsync` method because now you return a `CompletableFuture<T>` rather than `T`.

Reply:
Why do we need to use `table.refreshAsync().thenApply(__ -> table.get(key));` instead of 
```
 * table.refreshAsync().whenComplete((__, ex) -> {
 *     if (ex == null) {
 *         table.get(key);
 *     }
 * });
```

Reply:
A better example might be like:

```java
        CompletableFuture<MessageId> lastSendFuture = null;
        for (int i = 0; i < 10000; i++) {
            lastSendFuture = producer.newMessage().key("key-" + (i % 5)).value(("msg-" + i).getBytes()).sendAsync();
        }
        lastSendFuture.thenCompose(__ -> tableView.refreshAsync()).thenAccept(__ -> {
            final byte[] value = tableView.get("key-1");
            // ...
        });
```

i.e. all producers are controlled by yourself. You know there is no more message written to the topic, then it will be safe and meaningful to call `refreshAsync()` and other methods like `get()`.

Reply:
Take the example above, if you immediately call `refreshAsync` after sending 10000 messages rather than in the callback of `lastSendFuture`, the behavior is undefined.

Reply:
Why doesn't `whenComplete` work? And don't we need to handle exceptions?

Reply:
We can add `exceptionally` or `get`/`join` to handle the exception. `whenComplete` works but the use is not direct. Without the `refreshAsync`, the use of `TableView` is very simple and direct.

For example, assuming the application handles a request from client like:

```java
T handleQuery(Request req) {
    return tableView.get(req.key());
}
```

All the calls are synchronous so they are easy to use.

However, with a `whenComplete` callback, you can only process the key in the callback.

```java
T handleQuery(Request req) {
    var future = new CompletableFuture<T>();
    tableView.refreshAsync().whenComplete((__, e) -> {
        if (e != null) {
            future.complete(tableView.get(req.key()));
        }
    });
    return future.join(); // change it to synchronous
}
```

The code above is complicated, and it might be hard to users that get used to synchronous programming. For example, they might also write code like:

```java
T handleQuery(Request req) throws InterruptedException {
    var value = new AtomicReference<T>();
    tableView.refreshAsync().whenComplete((__, e) -> {
        if (e != null) {
            value.set(tableView.get(req.key()));
        }
    });
    while (value.get() == null)  {
        Thread.sleep(1);
    }
    return value.get();
}
```

In the previous use case, `thenApply` is more clear and simple.

```java
T handleQuery(Request req) {
    return tableView.thenApply(__ -> tableView.get(req.key()).join();
}
```

If any exception is thrown, `join()` (or `get()`) will throw it.

Reply:
Okay